teps/0122-reproducibility-of-complete-build-instructions.md

```diff
@@ -0,0 +1,297 @@
+      annotations: null
+```
+
+## Requirements
+- It must be possible to use the provenance generated by Tekton Chains for an artifact to create a bit by bit identical (best effort) artifact.
+  - Much of this will depend on the Task itself; note that the SLSA L4 required is "best effort".
+  - The provenance should contain all of the Task definition and runtime information required to reproduce the build.
+- The shas of the images run should continue to be included regardless (as this is part of the dependencies / build instructions)
+- It must be possible for users to construct policies based on the provenance that would allow them to determine if the build is ok - for example if they have policies around what pipeline tasks are acceptable to use, what parameters are acceptable, whether or not sidecars are allowed etc.
+  - According to [SLSA v0.2](https://slsa.dev/provenance/v0.2), if **invocation.configSource** is not available then **buildConfig** can be used to verify information about the build. If the policy requires validating the underlying task/pipeline spec that was not fetched from version control then the information from the spec required for validation must be embedded in the provenance’s buildConfig. 
+
+## Goals
+This proposal attempts to do the following:
+
+- Identify configuration, taskSpec and task-run-time information that lead to the production of the build but is currently missing in the provenance.
+- Identify where in the provenance to host this missing information.
+
+## Out-of-scope
+- Reproducibility of complete build instructions using a **Pipeline Run**.
+- A **Resolved Inputs** field for TaskRun where "resolved inputs" is a direct encoding of reproducibility (and verified during the build pod itself via the entrypointer). In the long term, it is useful to have reproducibility metadata built into pipelines so that the changes to the TaskRun will need to consider reproducibility and chains does not have to be constantly updated whenever there is a drift. This field will require its own TEP to spec-out the details and therefore it is currently out-of-scope.
+- **Level of reproducibility** in the provenance: 
+  - **Full** -- A user can rerun the exact steps with the exact inputs and get the exact results, with all <choose your favorite hash> signatures matching for inputs and artifacts. Until we have provenance that captures the full toolchain version stack -- possibly even down to the OS kernel version(?) -- we're going to be very hard-pressed to guarantee full reproducibility in the initial implementation.
+  - **Partial** -- We can guarantee that a user can run the exact steps with the exact inputs -- but the output artifacts could change due to toolchain changes. For example, imagine all inputs are identical but there is a new compiler optimization. The output artifact is functionally identical but the signature doesn't match.
```

Comment:
An easier example than "new compiler optimization" is "timestamps included in the output artifact". Those things get everywhere.

Reply:
Thanks! I will include that first.